File: mathematica/fun/test_list.m
(* test_list.m
                               wookay.noh at gmail.com
*)

assertEqual[expected_, got_] :=
  WriteString["stdout",
    If[expected == got,
      StringJoin["passed: ", ToString[expected], "\n"],
      StringJoin["Assertion failed\nExpected: ", ToString[expected],
        "\nGot: ", ToString[got], "\n"]]]


assertEqual[ {a,b,c} , List[a,b,c] ]

assertEqual[ a , Part[{a,b,c,d,e}, 1] ]
assertEqual[ a , {a,b,c,d,e}[[1]]     ]
assertEqual[ e , {a,b,c,d,e}[[-1]]    ]
assertEqual[ {a,b} , {a,b,c,d,e}[[{1,2}]] ]
assertEqual[ {b,c} , Part[{a,b,c,d,e}, 2;;3] ]
assertEqual[ {b,e,h} , Part[{{a, b, c}, {d, e, f}, {g, h, i}}, All, 2] ]

assertEqual[ {a,b,c} , Take[{a,b,c,d,e},3] ]
assertEqual[ {c,d,e} , Take[{a,b,c,d,e},-3] ]
assertEqual[ {b,c,d,e} , Rest[{a,b,c,d,e}] ]
assertEqual[ {b,c,d,e} , Drop[{a,b,c,d,e}, 1] ]
assertEqual[ {a,b,c,d} , Most[{a,b,c,d,e}] ]
assertEqual[ {a,b,c,d} , Drop[{a,b,c,d,e}, -1] ]

assertEqual[ {{2}} , Position[{a,b,c,d,e}, b] ]

assertEqual[ {3,4,4,3} , Select[{3,4,5,4,3}, # < 5 &] ]
assertEqual[ {3,4} , TakeWhile[{3,4,5,4,3}, # < 5 &] ]

assertEqual[ {a,c} , Pick[{a,b,c,d,e}, {1,0,1,0,0}, 1] ]
assertEqual[ {a,c} , Pick[{a,b,c,d,e}, {x,0,x,0,0}, x] ]

Exit[]

File: python/fun/test_list.py
def assert_equal(expected, got):
  if expected==got:
    print "passed: %s" % expected
  else:
    print "Assertion failed\nExpected: %s\nGot: %s" % (expected, got)

assert_equal( [2,3,4] , range(2,5) )
assert_equal( [0,1,2,3,4] , map(lambda x:x, range(0,5)) )
assert_equal( [2,3,0,1,6] , map(lambda x:x^2, range(0,5)) )
assert_equal( [4] , filter(lambda x:x>3, range(0,5)) )
assert_equal( [4] , filter(lambda x:x>3, range(0,5)) )

ary = [1,2,3]
assert_equal( None , list.append(ary, 4) )
assert_equal( [1,2,3,4] , ary)

ary = [1,2,3,3,3]
list.remove(ary,3)
assert_equal( [1,2,3,3] , ary)

ary = [1,2,3]
assert_equal( 3 , list.pop(ary) )
assert_equal( [1,2] , ary)

ary = [1,2,3,3,3]
assert_equal( 3 , list.count(ary, 3) )

ary = [1,2,3,3,3]
assert_equal( 2 , list.index(ary, 3) )

assert_equal( 10 , len(range(0, 10)))
assert_equal( 45 , reduce(lambda x,y:x+y, range(0, 10)) )

for v in [1,2,3]:
  pass

for iv in enumerate([1,2,3]):
  assert_equal( tuple, type(iv) )

for idx, v in enumerate([1,2,3]):
  assert_equal( v , idx+1 )

assert_equal([(0, 1), (1, 2), (2, 3)] , zip([0,1,2], [1,2,3]) )
